### PR TITLE
Counter battery fire

### PR DESCRIPTION
**Pull request by NickAragua**
Notable features:

- incoming off-board fire is checked to see if a friendly unit observes the hex that was hit. If it was, the firing unit can be targeted (and is called out in the round reports)
- adds an overlay to the targeting view that shows the user possible off-board targets for counter-battery fire (note that the weapon must be artillery and facing the right direction; off-board units can counter-battery fire other off-board units without restriction)
- implements processing for artillery attacks targeting a single entity by separating some logic from Server.artilleryDamageHex and putting it into a separate class.
- The bot can make counter-battery attacks too.

TODO: 
Actually have a proper icon instead of a transparent blue rectangle
Indicate a potential counter-battery target when one exists but the current unit can't fire on it, perhaps during the movement phase as well to help with planning

### CODE REVIEW

File: megamek/src/megamek/common/weapons/ArtilleryWeaponIndirectFireHandler.java [hunk 1]
```diff
@@ -455,6 +415,143 @@ else if ((null != bestSpotter) && !(this instanceof ArtilleryWeaponDirectFireHan
...
+            // in the ArtilleryWeaponDirectFireHandler
+            Coords origPos = targetPos;
+            int moF = toHit.getMoS();
+            if (ae.hasAbility("oblique_artillery")) {
+                // getMoS returns a negative MoF
+                // simple math is better so lets make it positive
+                if ((-moF -2) < 1) {
+                    moF = 0;
+                } else {
+                    moF = moF +2;
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/ArtilleryWeaponIndirectFireHandler.java]: Is this just `mof = Math.max(mof + 2, 0)` ?

Comment (NickAragua) [reply megamek/src/megamek/common/weapons/ArtilleryWeaponIndirectFireHandler.java]: I honestly have no idea, I just copy and pasted it from where it originally was.

**Comment by NickAragua** [comment]
Added the image; the "should we display the overlay" checks were being run 8x (once for each direction on the "should be visible" check, once for each direction when checking whether a particular off-board direction was being checked). Moved most of the checks so that they're only called once.

File: megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java [hunk 2]
```diff
@@ -0,0 +1,365 @@
...
+                (getCurrentGame().getPhase() == IGame.Phase.PHASE_TARGETING);
+        
+        if(!visible) {
+            return false;
+        }
+        
+        Mounted selectedArtilleryWeapon = clientgui.getBoardView().getSelectedArtilleryWeapon();
+        
+        // only relevant if we've got an artillery weapon selected for one of our own units
+        if(selectedArtilleryWeapon == null ) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java]:
```suggestion
        if(selectedArtilleryWeapon == null) {
```

File: megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java [hunk 3]
```diff
@@ -0,0 +1,365 @@
...
+            return false;
+        }
+        
+        // the artillery weapon needs to be using non-homing ammo
+        Mounted ammo = selectedArtilleryWeapon.getLinked();
+        AmmoType ammoType = (AmmoType) ammo.getType();
+        if((ammoType.getMunitionType() == AmmoType.M_HOMING) 
+                && ammo.curMode().equals("Homing")) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java]: Do we have constants/statics for these modes?

File: megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java [hunk 4]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    @Override
+    public boolean isHit(Point point, Dimension size) {
+        Point actualPoint = point;
+        actualPoint.x = (int) (point.getX() + clientgui.getBoardView().getDisplayablesRect().getX());
+        actualPoint.y = (int) (point.getY() + clientgui.getBoardView().getDisplayablesRect().getY());
+        
+        for(OffBoardDirection direction : OffBoardDirection.values()) {
+            if(direction != OffBoardDirection.NONE) {
```
Comment (sixlettervariables) [inline megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java]: If `buttons` will never contain `OffBoardDirection.None` you can probably omit this if test.

File: megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        case WEST:
+            xPosition = boundingRectangle.x + EDGE_OFFSET;
+            yPosition = boundingRectangle.y + (int) (boundingRectangle.height / 2) - (int) (WIDE_EDGE_SIZE / 2);
+            return new Rectangle(xPosition, yPosition, WIDE_EDGE_SIZE, NARROW_EDGE_SIZE); // used to be NARROW_EDGE_SIZE, WIDE_EDGE_SIZE);
+        // south rectangle is wider than narrower, and at the bottom of the board view
+        case SOUTH:
+            xPosition = boundingRectangle.x + (int) (boundingRectangle.width / 2) - (int) (WIDE_EDGE_SIZE / 2);
+            yPosition = boundingRectangle.y + boundingRectangle.height - EDGE_OFFSET - NARROW_EDGE_SIZE;
+            return new Rectangle(xPosition, yPosition, WIDE_EDGE_SIZE, NARROW_EDGE_SIZE);
+        /// east rectangle is narrower than wider, and at the right of the board view, but to the left of the unit overview panel
```
Comment (sixlettervariables) [inline megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java]:
```suggestion
        // east rectangle is narrower than wider, and at the right of the board view, but to the left of the unit overview panel
```

File: megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+    }
+
+    @Override
+    public boolean slide() {
+        return false;
+    }
+    
+    /**
+     * Worker function that handles a click on a 'counterbattery fire' overlay button.
+     * Possible shows a target selection popup
```
Comment (sixlettervariables) [inline megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java]:
```suggestion
     * Possibly shows a target selection popup
```

File: megamek/src/megamek/common/weapons/AreaEffectHelper.java [hunk 7]
```diff
@@ -67,6 +81,46 @@ private static void initializeFuelAirBlastRadiusIndexData() {
...
+    private static void initializeNukeStats() {
+        nukeStats = new HashMap<>();
+        
+        NukeStats nukeEntry = new NukeStats();
+        nukeEntry.baseDamage = 100;
+        nukeEntry.degradation = 5;
+        nukeEntry.secondaryRadius = 40;
+        nukeEntry.craterDepth = 0;
+        
+        nukeStats.put(0, nukeEntry);
```
Comment (sixlettervariables) [inline megamek/src/megamek/common/weapons/AreaEffectHelper.java]: Probably worth a comment explaining how this `nukeStats` map works.

File: megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java (hunk 3 above)
Comment (NickAragua) [reply megamek/src/megamek/client/ui/swing/OffBoardTargetOverlay.java]: Not at this time. There are quite a few places where this comparison is made, so I think I'm going to hold off on it for this PR, for the sake of minimizing its impact.

**Comment by NickAragua** [comment]
Made most of the code review changes. Will hold off on the "homing" refactor for now.

**Comment by NickAragua** [comment]
Never mind, it wasn't that wide-spread. Also fixed a bug where units being targeted by counter-battery fire wouldn't report observation of the incoming shot for counter-battery fire purposes, and would incorrectly "observe" on-board units.

**Pull request merged**